src/qibo/backends/tensorflow.py

```diff
@@ -306,6 +306,16 @@ def entanglement_entropy(self, rho):
         entropy = self.np.sum(masked_eigvals * spectrum) / self.np.log(2.0)
         return entropy, spectrum
 
+    def calculate_norm(self, state, order=2):
+        state = self.cast(state)
+        return self.tf.norm(state, ord=order)
+
+    def calculate_norm_density_matrix(self, state, order="nuc"):
+        state = self.cast(state)
+        if order == "nuc":
+            return self.np.trace(state)
```

Comment:
Regarding the discussion on using tensorflow primitives, indeed, if someone uses gradient descent over a numpy operation it will break the gradient calculation. However, note that in this case `self.np` is not `numpy` but
```py
import tensorflow.experimental.numpy as tnp
```
(see the `__init__`). Back at the time we were writing this backend this was an easy way to port numpy operations to tensorflow, but I am not sure what is the status now. I am also not sure how it will behave in terms of the gradient.

Reply:
Indeed! 
This works well.